src/sqlfluff/utils/reflow/classes.py

```diff
@@ -0,0 +1,283 @@
+    root_segment: BaseSegment
+
+    def __post_init__(self):
+        """Validate integrity"""
+        assert self.elements
+        # Check root segment is shared
+        assert all(elem.root_segment is self.root_segment for elem in self.elements)
+        # Check first and last
+        try:
+            if isinstance(self.elements[0], ReflowPoint):
+                # Check odds are all points
+                assert all(isinstance(elem, ReflowPoint) for elem in self.elements[::2])
+                # Check evens are all blocks
+                assert all(
+                    isinstance(elem, ReflowBlock) for elem in self.elements[1::2]
+                )
+            else:
+                # Check odds are all points
+                assert all(isinstance(elem, ReflowBlock) for elem in self.elements[::2])
+                # Check evens are all blocks
+                assert all(
+                    isinstance(elem, ReflowPoint) for elem in self.elements[1::2]
+                )
+        except AssertionError as err:
+            reflow_logger.exception(
+                "Assertion check on ReflowSequence failed. " "Elements: %s",
+                self.elements,
+            )
+            raise err
+
+    @classmethod
+    def from_raw_segments(
+        cls, segments: Sequence[RawSegment], root_segment: BaseSegment
+    ):
+        """Construct a ReflowSequence from a sequence of raw segments.
+
+        Aimed to be the basic constructor, which other more specific
+        ones may fall back to.
+        """
+        elem_buff = []
+        seg_buff = []
+        SeqClass: Union[None, Type[ReflowBlock], Type[ReflowPoint]] = None
+        for seg in segments:
+            NextClass: Union[Type[ReflowBlock], Type[ReflowPoint]]
+            # NOTE: Can indents be in a reflow block? I assume not.
+            # Placeholders certainly not because they also get indented.
+            if seg.is_type("whitespace", "newline", "end_of_file"):
+                NextClass = ReflowPoint
+            else:
+                NextClass = ReflowBlock
+
+            reflow_logger.debug(
+                "    Evaluating %s. NextClass: %s, SeqClass: %s",
+                seg,
+                NextClass.__name__,
+                SeqClass.__name__ if SeqClass else None,
+            )
+            if not SeqClass:
+                # Extend the buffer if we're the first segment.
+                seg_buff.append(seg)
+            elif SeqClass is ReflowPoint and NextClass is ReflowPoint:
+                # Extend the buffer if we're still in s spacing section.
+                seg_buff.append(seg)
+            elif SeqClass is ReflowBlock and (
+                seg.is_type("indent") or all(s.is_type("indent") for s in seg_buff)
+            ):
+                # Extend the buffer if this is an indent
+                # OR the existing block is all indent.
+                seg_buff.append(seg)
+            else:
```

Comment:
These if else block and the fall through (`SeqClass = NextClass`) are hard to read 
```suggestion
            if _should_extend_buffer(seg, SeqClass, NextClass):
                seg_buff.append(seg)
                SeqClass = NextClass
                continue

```

Reply:
Yeah I agree - rather than make a new function though I'm going to keep the conditions in the function, but structure it more like what you've put here.